### PR TITLE
Support asymmetric signature algorithm for issuing id token

### LINKED ISSUE #4155: Support multiple asymmetric signature algorithms. (more details)

### PR DESCRIPTION
**Pull request by wansir**
Signed-off-by: hongming <hongming@kubesphere.io>

### What type of PR is this?

/kind feature

### What this PR does / why we need it:

Currently, we issue tokens by the HMAC-SHA256 algorithm. Symmetric signatures MUST NOT be used by public (non-confidential) clients because of their inability to keep secrets. For security reasons, we should use the asymmetric algorithm sign `id_token` by default. 

### Which issue(s) this PR fixes:

Fixes #4155

### Does this PR introduced a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/apiserver/authentication/oauth/error.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2020 The KubeSphere Authors.
```
Comment (LinuxSuRen) [inline pkg/apiserver/authentication/oauth/error.go:2]: 2020 or 2021?

Comment (wansir) [reply pkg/apiserver/authentication/oauth/error.go]: fixed

**Comment by zryfish** [comment]
This PR shares a lot of code with PR #3525, should merge #3525 first.

**Comment by wansir** [comment]
> This PR shares a lot of code with PR #3525, should merge #3525 first.

#3525 depends on this PR, pls see commit logs.  We should merge this PR first.

**Comment by wansir** [comment]
All of the commits are included in #3525
/close

**Pull request closed without merging**